### PR TITLE
make convert_image_dtype scriptable

### PR DESCRIPTION
**Pull request by nairbv**
I'm not sure if this is a good idea since it might be slower, but wanted to see if I could make convert_image_dtype scriptable.

iinfo isn't supported by torchscript: https://github.com/pytorch/pytorch/issues/41492

This PR also moves convert_image_dtype to functional_tensor since it only applies to tensor ops, and because functional_tensor can't import from functional.

### CODE REVIEW

File: torchvision/transforms/functional.py [hunk 1]
```diff
@@ -124,7 +124,24 @@ def pil_to_tensor(pic):
     return img
 
 
-def convert_image_dtype(image: torch.Tensor, dtype: torch.dtype = torch.float) -> torch.Tensor:
+# torch.iinfo isn't scriptable so using this helper function
+# https://github.com/pytorch/pytorch/issues/41492
+def _max_value(dtype: int) -> int:
```
Comment (nairbv) [inline torchvision/transforms/functional.py:129]: it might be possible to memoize this if performance is an issue, or just convert to a lookup.

File: torchvision/transforms/functional.py [hunk 2]
```diff
@@ -124,7 +124,24 @@ def pil_to_tensor(pic):
...
+    bits = 1
+    max_value = torch.tensor(-signed, dtype=torch.long)
+    while(True):
+        next_value = a.pow(bits - signed).sub(1)
+        if next_value > max_value:
+            max_value = next_value
+            bits *= 2
+        else:
+            return max_value.item()
+    return max_value.item()
```
Comment (nairbv) [inline torchvision/transforms/functional.py:141]: jit didn't like the annotation without this, even though it's unreachable.

File: torchvision/transforms/functional.py [hunk 3]
```diff
@@ -124,7 +124,24 @@ def pil_to_tensor(pic):
...
+
+
+def convert_image_dtype(image: torch.Tensor, dtype: int = torch.float) -> torch.Tensor:
```
Comment (nairbv) [inline torchvision/transforms/functional.py:144]: jit doesn't support torch.dtype but defines apis that view it as an int

File: torchvision/transforms/functional_tensor.py [hunk 4]
```diff
@@ -228,13 +310,11 @@ def adjust_gamma(img: Tensor, gamma: float, gain: float = 1) -> Tensor:
     result = img
     dtype = img.dtype
     if not torch.is_floating_point(img):
-        result = result / 255.0
+        result = convert_image_dtype(result, torch.float32)
```
Comment (nairbv) [inline torchvision/transforms/functional_tensor.py:313]: unfortunately torch.get_default_dtype() isn't scriptable either.

**Comment by nairbv** [comment]
> @nairbv I've already started this in #2313. If we want to do this now, maybe we can "merge" the PRs.

ah, I should have checked first. I was curious if I could and didn't realize someone was working on it already.

File: torchvision/transforms/functional_tensor.py (hunk 4 above)
Comment (fmassa) [reply torchvision/transforms/functional_tensor.py:313]: It's fine to default to `float32` for all computations, I think that using `get_default_dtype()` could lead to surprises if the user switched the defaults

File: torchvision/transforms/functional_tensor.py [hunk 5]
```diff
@@ -18,6 +18,88 @@ def _get_image_size(img: Tensor) -> List[int]:
...
+    bits = 1
+    max_value = torch.tensor(-signed, dtype=torch.long)
+    while(True):
+        next_value = a.pow(bits - signed).sub(1)
+        if next_value > max_value:
+            max_value = next_value
+            bits *= 2
+        else:
+            return max_value.item()
+    return max_value.item()
```
Comment (fmassa) [inline torchvision/transforms/functional_tensor.py:35]: Nice! Although I wonder if simply having a table with the min / max values wouldn't be a simpler (and more efficient) workaround while `iinfo` isn't implemented?

File: torchvision/transforms/functional_tensor.py [hunk 6]
```diff
@@ -18,6 +18,88 @@ def _get_image_size(img: Tensor) -> List[int]:
...
+
+
+def convert_image_dtype(image: torch.Tensor, dtype: int = torch.float) -> torch.Tensor:
```
Comment (fmassa) [inline torchvision/transforms/functional_tensor.py:38]: Can you move this function back to `functional.py`? My thinking about `functional_pil` and `functional_tensor` is that those are implementation details, and the user should only worry about `functional.py`.

File: torchvision/transforms/functional_tensor.py (hunk 5 above)
Comment (nairbv) [reply torchvision/transforms/functional_tensor.py:35]: It would certainly be more efficient, just slightly more work to maintain. I'm open to either way.

File: torchvision/transforms/functional_tensor.py (hunk 6 above)
Comment (nairbv) [reply torchvision/transforms/functional_tensor.py:38]: The issue that raises is then the implementation of `adjust_gamma` in functional_tensor (and any other functional impl's in which we'd like to use this) would need to depend on functional.py, but jit doesn't like that import.

Another option could be to have a `convert_image_dtype` in `functional.py` that takes either a PIL or Tensor, and delegates to this implementation as do other transforms. That doesn't seem to make sense to me unless there's also an equivalent PIL-based convert_image_dtype transform we would want to implement/use somewhere. To me this is seeming like a tensor-specific helper transform.

Comment (fmassa) [reply torchvision/transforms/functional_tensor.py:38]: Good point about the cyclic import. I don't think we want an implementation for PIL, but here is one proposal: add in `functional.py` something like

```python
# import to main namespace
# this is temporary until we merge the implementation of
# F_t inside functional.py
convert_image_dtype = F_t.convert_image_dtype
```

File: torchvision/transforms/functional_tensor.py (hunk 5 above)
Comment (fmassa) [reply torchvision/transforms/functional_tensor.py:35]: Given that you already have this implementation here, let's keep it as is for now and add a TODO saying to replace it with `torch.iinfo` when it gets torchscript support.

**Comment by pmeier** [comment]
@fmassa Do you want me to close #2313 since we are moving forward with this PR?

**Comment by fmassa** [comment]
@pmeier yes, let's move forward with this PR, thanks a lot for working on that other PR!

File: torchvision/transforms/transforms.py [hunk 7]
```diff
@@ -131,7 +131,7 @@ def __init__(self, dtype: torch.dtype) -> None:
         self.dtype = dtype
 
     def __call__(self, image: torch.Tensor) -> torch.Tensor:
-        return F.convert_image_dtype(image, self.dtype)
+        return F_t.convert_image_dtype(image, self.dtype)
```
Comment (fmassa) [inline torchvision/transforms/transforms.py:134]: Can we revert the changes in this file, given that we have now exposed `convert_image_dtype` to `functional.py`?
This way `transforms.py` only depends on `functional.py`

File: test/test_functional_tensor.py [hunk 8]
```diff
@@ -323,7 +323,7 @@ def test_adjust_gamma(self):
 
                 rbg_tensor = adjusted_tensor
                 if adjusted_tensor.dtype != torch.uint8:
-                    rbg_tensor = F.convert_image_dtype(adjusted_tensor, torch.uint8)
+                    rbg_tensor = F_t.convert_image_dtype(adjusted_tensor, torch.uint8)
```
Comment (fmassa) [inline test/test_functional_tensor.py:326]: I wonder if we should revert those changes now that we have the implementation back in `functional.py` ?

File: torchvision/transforms/functional_tensor.py [hunk 9]
```diff
@@ -18,6 +18,88 @@ def _get_image_size(img: Tensor) -> List[int]:
...
+    Raises:
+        RuntimeError: When trying to cast :class:`torch.float32` to :class:`torch.int32` or :class:`torch.int64` as
+            well as for trying to cast :class:`torch.float64` to :class:`torch.int64`. These conversions might lead to
+            overflow errors since the floating point ``dtype`` cannot store consecutive integers over the whole range
+            of the integer ``dtype``.
+    """
+    if image.dtype == dtype:
+        return image
+
+    if torch.empty(0, dtype=image.dtype).is_floating_point():
```
Comment (fmassa) [inline torchvision/transforms/functional_tensor.py:62]: nit: can you add a comment saying something like 
```python
# TODO: replace with image.dtype.is_floating_point when torchscript supports it
```

File: torchvision/transforms/functional_tensor.py (hunk 6 above)
Comment (vfdev-5) [reply torchvision/transforms/functional_tensor.py]: If used defined `convert_image_dtype = F_t.convert_image_dtype` inside `functional.py`, then docstring for the method is not rendered. I moved back the implementation with the only dispatch to `F_t`.

**Comment by nairbv** [comment]
Thanks @pmeier and @vfdev-5!
